### PR TITLE
feat(minor-ampd): check if configured fee denomination can be paid at startup

### PR DESCRIPTION
**Pull request by cgorenflo**
AXE-4206

### CODE REVIEW

File: ampd/src/broadcaster/queries.rs
Comment (fish-sammy) [inline ampd/src/broadcaster/queries.rs:1]: Personally, I'd wanna combine this file with client.rs.

**Review by fish-sammy — APPROVED** [review]

File: ampd/src/broadcaster/clients.rs [hunk 1]
```diff
@@ -59,14 +61,37 @@ pub trait AccountQueryClient {
...
     async fn account(
         &mut self,
         request: QueryAccountRequest,
     ) -> Result<QueryAccountResponse, Status> {
-        self.account(request)
+        return self
+            .account(request)
             .await
             .map(Response::into_inner)
+            .map_err(Report::from);
```
[the remark this answers is not in the document]
Comment (cgorenflo) [reply ampd/src/broadcaster/clients.rs:73]: I've recently worked with languages other than rust, it's just a slipup

File: ampd/src/broadcaster/clients.rs
[the remark this answers is not in the document]
Comment (cgorenflo) [reply ampd/src/broadcaster/clients.rs:1]: but those things are only visible to the broadcaster itself and not supposed to be used by anyone else, that's why I created the `queries` submodule

File: ampd/src/broadcaster/mod.rs [hunk 2]
```diff
@@ -176,11 +246,12 @@ where
 
 impl<T, S, Q> BroadcastClient<T, S, Q>
 where
-    T: clients::BroadcastClient,
-    Q: clients::AccountQueryClient,
+    T: clients::BroadcastClient + Send,
+    S: Multisig + Send + Sync,
```
[the remark this answers is not in the document]
Comment (cgorenflo) [reply ampd/src/broadcaster/mod.rs:250]: I'm pretty sure that came from your changes, but I'll check again

File: ampd/src/broadcaster/clients.rs
[the remark this answers is not in the document]
Comment (cgorenflo) [reply ampd/src/broadcaster/clients.rs:1]: refactored the file structure

File: ampd/src/broadcaster/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cgorenflo) [reply ampd/src/broadcaster/mod.rs:286]: these constraints are needed to implement the `Broadcaster` trait, and the struct is pretty useless without that, so I opted to make those constraints global for this struct everywhere

File: ampd/src/broadcaster/queries.rs
Comment (cgorenflo) [reply ampd/src/broadcaster/queries.rs:1]: I reshuffled the file structure

File: ampd/src/broadcaster/mod.rs [hunk 3]
```diff
@@ -88,30 +104,120 @@ impl Default for Config {
...
+    T: cosmos::BroadcastClient + Send,
+    S: Multisig + Send + Sync,
+    A: cosmos::AccountQueryClient + Send,
+    B: cosmos::BalanceQueryClient,
+{
+    pub async fn validate_fee_denomination(mut self) -> Result<BasicBroadcaster<T, S, A>, Error> {
+        let denom: Denom = self.config.gas_price.denom.clone().into();
+        let address: TMAddress = self.derive_address()?;
+
+        ensure!(
```
Comment (fish-sammy) [inline ampd/src/broadcaster/mod.rs:141]: Cool! Didn't know this.

File: ampd/src/broadcaster/mod.rs [hunk 4]
```diff
@@ -88,30 +104,120 @@ impl Default for Config {
...
+        )
+    }
+}
+
+fn extract_non_zero_amount(coin: Coin) -> Option<Amount> {
+    Some(coin.amount).filter(|amount| !amount.is_zero())
+}
+
+#[derive(Debug)]
+pub struct BasicBroadcaster<T, S, Q>
```
Comment (fish-sammy) [inline ampd/src/broadcaster/mod.rs:198]: Calling it `BasicBroadcaster` just so that it's name won't conflict with the trait?

**Review by fish-sammy — APPROVED** [review]

File: ampd/src/broadcaster/cosmos.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
+//! Abstractions that decouple the broadcaster from the explicit clients of the cosmrs crate.
+//!
+//! In this module, new traits are defined and implemented for several cosmrs clients,
+//! so the broadcaster only needs to depend on the traits that it owns itself. This makes it less
+//! vulnerable to unexpected changes in the external crate and helps with unit testing,
+//! because the client traits also provide auto-mocks.
```
Comment (fish-sammy) [inline ampd/src/broadcaster/cosmos.rs:6]:
```suggestion
//! because the client traits also provide auto-mocks.
```

**Pull request merged**